core/src/main/java/tech/pegasys/web3signer/core/Runner.java

```diff
@@ -224,22 +224,45 @@ private void registerHttpHostAllowListHandler(final OpenAPI3RouterFactory openAp
   }
 
   private void registerSwaggerUIRoute(final Router router) throws IOException {
-    final URL indexResourceUrl = Resources.getResource(OPENAPI_INDEX_RESOURCE);
-    final URL openApiSpecUrl = Resources.getResource(getOpenApiSpecResource());
-    final String indexHtml = Resources.toString(indexResourceUrl, Charsets.UTF_8);
-    final String openApiSpecYaml = Resources.toString(openApiSpecUrl, Charsets.UTF_8);
-
-    router
-        .route(HttpMethod.GET, SWAGGER_ENDPOINT + "/web3signer.yaml")
-        .produces(CONTENT_TYPE_YAML)
-        .handler(ResponseContentTypeHandler.create())
-        .handler(routingContext -> routingContext.response().end(openApiSpecYaml));
-
-    router
-        .routeWithRegex(HttpMethod.GET, SWAGGER_ENDPOINT + "|" + SWAGGER_ENDPOINT + "/*")
-        .produces(CONTENT_TYPE_TEXT_HTML)
-        .handler(ResponseContentTypeHandler.create())
-        .handler(routingContext -> routingContext.response().end(indexHtml));
+    LOG.info(" Registering /swagger-ui routes...");
+    /* vertx static handler doesn't seem to work with OpenApi3Router. So we manually load all the
+     resources under /openapi, fix relative schema links and register them.
+    */
+    try (final Stream<URL> webrootYamlFiles = Resources.find("/openapi**.*")) {
+      webrootYamlFiles.forEach(
+          yaml -> {
+            final String path = yaml.getPath();
+            // the path is in file://.jar!<path> format. We need the non-jar path.
+            final String yamlRoute = StringUtils.substringAfter(path, "!/openapi");
+            LOG.info("Registering: {}/{}", SWAGGER_ENDPOINT, yamlRoute);
+            final String file;
+            try {
+              file = com.google.common.io.Resources.toString(yaml, StandardCharsets.UTF_8);
+            } catch (final IOException e) {
+              throw new UncheckedIOException(e);
+            }
+
+            // adjust the ref (which works with OpenApi3Router) to relative that works with
+            // swagger-ui
+            final String adjustedFile =
+                StringUtils.replace(file, "/openapi/eth2/signing/schemas.yaml", "../schemas.yaml");
```

Comment:
quite annoying to have to do this twice, both in the gradle task and here :( 

Once I add the new API endpoints I'll have another path to replace too on top of this one. Will be `/openapi/eth2/keymanager/schemas.yaml`. 

Would be great to make this replace logic generic so I don't have to touch this code once I add the new endpoints. Maybe with a regex? If starts with `/` -> replace everything before the last slash with `..` or something like that.

Reply:
Also realized that this flattening of files into one folder means that all filenames must be unique right? Kind of a bummer. Anyway to preserve some of the folder hierarchy and still use relative paths? Like `../signing/schemas.yaml` and `../keymanager/schemas.yaml`

Would also make the string replace logic simpler that way. Would just be replace `/openapi/eth2` with `..` and done

Reply:
@joaquim-verges 

I am about to re-do this logic slightly ... apparently the Vertx's OpenApi3Router (which internally uses swagger-parser) is not fool-proof (i.e. fails to parse) when it deals with relative paths in `$ref` (such as sign.yaml referencing `../schema.yaml#....` where it sits two folder deep. It works if all files are on same level. On the other hand, tools like swagger-ui and redoc (which are javascript based) works fine with such relative paths.

I am adding a bit of logic to modify the relative `$ref` and `discriminator` entries to absolute paths for OpenApi3Router handling.

The end file structure that should work would look like:
~~~
openapi
├── eth1
│   └── web3signer.yaml
├── eth2
│   ├── signing
│   │   ├── paths
│   │   │   ├── public_keys.yaml
│   │   │   ├── reload.yaml
│   │   │   ├── sign.yaml
│   │   │   └── upcheck.yaml
│   │   └── schemas.yaml
│   └── web3signer.yaml
├── filecoin
│   └── web3signer.yaml
└── index.html
~~~

Reply:
file structure looks good to me! Def had a hard time making relative paths work with Vertx, if there's a robust way to make it work that would be the cleanest for sure.